asteroid/engine/schedulers.py

```diff
@@ -0,0 +1,69 @@
+
+class _BaseScheduler(object):
+    def __init__(self, optimizer):
+        self.optimizer = optimizer
+        self.step_num = 0
+
+    def zero_grad(self):
+        self.optimizer.zero_grad()
+
+    def _get_lr(self):
+        raise NotImplementedError
+
+    def _set_lr(self, lr):
+        for param_group in self.optimizer.param_groups:
+            param_group['lr'] = lr
+
+    def step(self):
+        self.step_num += 1
+        lr = self._get_lr()
+        self._set_lr(lr)
+        self.optimizer.step()
+
+    def load_state_dict(self, state_dict):
+        self.__dict__.update(state_dict)
+
+    def state_dict(self):
+        return {key: value for key, value in self.__dict__.items()}
+
+
+
+class NoamScheduler(_BaseScheduler):
+    def __init__(self, optimizer, d_model, warmup_steps, scale=1.0):
```

Comment:
`d_model` is the projection dim of the MHA?

Reply:
the input dimension of MHA. In our case input and projection dim are the same